Add support for sticker packs, guild stickers, sticker routes

drklee3 opened the pull request:
### Description

Adds the following features:

* Add/modify/delete guild stickers
* Fetch sticker information
* Fetch list of all nitro stickers
* Sending stickers in message (up to 3)

Breaking change as sticker module is moved out of the channel module and to the root model module.

This also adds an additional `mime` and `mime_guess` dependencies as Discord requires the mime type on the file part when creating a sticker via multipart/form-data requests otherwise it will return a 500 Internal Server Error. This dependency is also used by `reqwest` internally.

#### Relevant documentation:

* https://github.com/discord/discord-api-docs/pull/3127
* https://github.com/discord/discord-api-docs/pull/3128
* https://github.com/discord/discord-api-docs/pull/3171

### Tested

- Sending / receiving stickers in messages
- Get / create / edit / delete sticker

nickelc committed:
Update `async-tungstenite` to 0.12 (#1181)

arqunis committed:
Conditionally instantiate `WebSocketConfig` to fix tokio 0.2 compilation (#1214)

arqunis committed:
Make `RichInvite::inviter` optional and replace `InviteUser` with `User` (#1213)

nickelc committed:
Update `async-tungstenite` to 0.13 (#1228)

Lakelezz committed:
Remove `Box` indirection on `framework`. (#1240)

This commit removes the `Box` around the framework, allowing to `as`-cast for the `ClientBuilder::framework_arc` method.
The `as`-cast is now performed in the method itself, accepting a generic framework instead, easing the use of the API.

Licenser committed:
Add support for simd-json via a feature flag (#1237)

This adds support for parsing json with the `simd-json` crate for increased performance.
However, it must be enabled with the `simd-json` feature flag, as not all processors support
all SIMD instructions. In addition, the minimum-supported Rust version has been increased to 1.49.

kangalio committed:
Add the name of the command to the `Dispatch` error hook (#1344)

Closes #876. 

The dispatch error hook now receives a fourth parameter with the command name.

kangalio committed:
Alter deserialisation of `Presence::user` to properly handle optional fields (#1361)

hybras committed:
Ensure `Activity::url` is a valid URL at a type-level (#1373)

kangalio committed:
Add support for adding attachments in message edits (#1378)

This commit makes it possible to add new attachments to a message when editing it.

This commit:
- creates a new method `Http::edit_message_and_attachments` (that method is to `edit_message` what `send_files` is to `send_message`);
  - there's a lot of duplication between that new method and `Http::edit_message_and_attachments` and `Http::send_files`, but I don't know enough about serenity's http design to improve the situation;
- changes calls of `Http::edit_message` to `Http::edit_message_and_attachments` where appropriate;
- creates a new method `EditMessage::attachment` to add a new attachment;
- adds a new field to `EditMessage` which stores new attachments (breaking change); 
- adds a lifetime parameter to `EditMessage` (breaking change);
- moves the `AttachmentType` serialization into an HTTP form part into a function to avoid code duplication.

drklee3 committed:
Update sticker fields

drklee3 committed:
Make `Embed::colour` an `Option` (#1392)

Prevents changing embed colours to black when editing a message.

When an embed colour isn't provided when creating a message, the API returns messages with colour as undefined for default colour. The colour field has the `#[serde(default)]`  attribute, which changes that into a `0` instead.

drklee3 committed:
feat(http,model): Add guild sticker support

drklee3 committed:
fix(http): Fix json methods

drklee3 committed:
Add cover_sticker and banner_url StickerPack methods

drklee3 committed:
Add sticker_items message field

Zaitara commented on src/http/client.rs:
```diff
@@ -688,6 +688,81 @@ impl Http {
+            },
+            AttachmentType::File {
+                file,
+                filename,
+            } => {
+                let mut buf = Vec::new();
+                file.try_clone().await?.read_to_end(&mut buf).await?;
+
+                multipart = multipart.part("file", Part::stream(buf).file_name(filename));
+            },
+            AttachmentType::Path(path) => {
+                let filename =
+                    path.file_name().map(|filename| filename.to_string_lossy().into_owned());
+                let mut file = File::open(path).await?;
+                let mut buf = vec![];
+                file.read_to_end(&mut buf).await?;
+
+                let part = match filename {
+                    Some(filename) => Part::bytes(buf).file_name(filename),
+                    None => Part::bytes(buf),
+                };
+
+                multipart = multipart.part("file", part);
+            },
+            AttachmentType::Image(url) => {
+                let url = Url::parse(url).map_err(|_| Error::Url(url.to_string()))?;
+                let filename = url
+                    .path_segments()
+                    .and_then(|segments| segments.last().map(ToString::to_string))
+                    .ok_or_else(|| Error::Url(url.to_string()))?;
+                let response = self.client.get(url).send().await?;
+                let mut bytes = response.bytes().await?;
+                let mut picture: Vec<u8> = vec![0; bytes.len()];
+                bytes.copy_to_slice(&mut picture[..]);
+                multipart =
+                    multipart.part("file", Part::bytes(picture).file_name(filename.to_string()));
+            },
+        }
+
+        multipart = multipart.text("payload_json", serde_json::to_string(&map)?);
```
This code seems to be repeated a lot in the library

drklee3 replied:
True. I was thinking if I can get #1269 fixed and removing the duplicate code with that fix since it's related, so it will probably be in a separate PR.

drklee3 committed:
Remove deprecated Message::stickers field

tylerd008 committed:
Implement optional pagination for `Http::get_guilds` (#1403)

arqunis committed:
Change return signatures of help commands to return `Result`s (#1412)

This changes the return type of the default help commands from `Option` to `Result` to allow the caller to decide how they want to present the error that occurred when sending the response.

drklee3 committed:
Correctly cache guild categories on add/delete/update (#1405)

This deserializes `Guild` channels from `GuildCreate` as the `Channel` enum instead of `GuildChannel`. This also prevents the merging of guild channels and categories in `Cache::channels` when they should be separate.

This also correctly updates the guild channels in the cache (e.g. when using `Message::guild(ctx)`) instead of *only* `Cache::categories`.

Additionally, this is a breaking change as `Guild::channels` has been changed from `HashMap<ChannelId, GuildChannel>` to `HashMap<ChannelId, Channel>`.

rasm47 committed:
Remove `Shard::is_shutdown` (#1414)

This commit removes the `Shard::shutdown` field and the `Shard::is_shutdown` method. The method always returned false as `Shard::shutdown` was never modified anywhere, which left the method useless.

arqunis committed:
Configure project to be always compiled with native CPU support (#1416)

This adds a Cargo configuration file that defines flags for every invocation of `rustc` on Serenity and its dependencies. This removes the need to explicitly specify those flags when building the library, especially when testing `simd-json` code.

Fixes #1415

drklee3 committed:
Update new permissions, Guild::sticker field

drklee3 committed:
fix: Use new correct permission name

drklee3 committed:
fix doc link to new perm

Zaitara commented on src/model/sticker.rs:
```diff
@@ -0,0 +1,88 @@
+    /// For guild stickers, a unicode emoji representing the sticker's
+    /// expression. For nitro stickers, a comma-separated list of related
+    /// expressions.
+    pub tags: String,
```
You should make it an array to keep consistency and to make it easier to handle -which would mean that the array would be of length 1 if it is a guild owned one-

drklee3 committed:
Make Sticker::description Option

drklee3 committed:
Deserialize sticker tags as vec

drklee3 committed:
Serde default message::sticker_items

drklee3 committed:
Add list sticker packs, get sticker routes

drklee3 committed:
Add create/edit/delete StickerId/GuildId/Guild/PartialGuild methods

drklee3 committed:
Add Sticker methods

drklee3 committed:
Add errors section to StickerItem::to_sticker

drklee3 committed:
Remove unused imports

drklee3 committed:
Remove extra sticker builder doc links

drklee3 committed:
cargo fmt

drklee3 committed:
Fix doc links

drklee3 committed:
Add sticker gateway event

Zaitara commented on src/model/utils.rs:
```diff
@@ -290,6 +290,48 @@ pub fn serialize_roles<S: Serializer>(
+    vec: &Vec<String>,
+    serializer: S,
+) -> StdResult<S::Ok, S::Error> {
+    serializer.serialize_str(&vec.join(" "))
```
```suggestion
    serializer.serialize_str(&vec.join(","))
```
To be consistent with the deserialization

drklee3 replied:
Oh yeah oops, thanks for the catch.

drklee3 committed:
Split by ', ' when (de)serializing comma separated strings

drklee3 committed:
Integrate attachments into the `Request` system (#1408)

This bakes multipart forms into `Request`s when making HTTP calls to Discord's API for sending attachments. As a result, this standardises the behaviour for endpoints that allow you to send attachments to use the same ratelimiting code as every other endpoint, rather than those endpoints performing raw requests and bypassing the ratelimiter.

Fixes #1269

drklee3 committed:
Remove duplicated multipart sticker_create code

drklee3 committed:
Remove AttachmentType::into_http_form_part

sry kangalioo i delete your code

drklee3 committed:
Rename MANAGE_EMOJI_AND_STICKERS to MANAGE_EMOJIS_AND_STICKERS

drklee3 committed:
Update docs and limits

drklee3 committed:
Add sticker audit log events

drklee3 committed:
Add sticker_id methods to CreateMessage

drklee3 committed:
One more rename

drklee3 committed:
Merge branch 'next' into guild-stickers

drklee3 committed:
Check sticker lengths < 3

drklee3 committed:
Add Sticker::image_url

drklee3 committed:
Consolidate message length checking methods

drklee3 committed:
Sticker::available doc guild clarification

drklee3 committed:
cargo fmt/clippy

drklee3 committed:
Make CreateMessage sticker methods consistent with multi embed methods

drklee3 committed:
fix: Use multipart fields for sticker params

drklee3 committed:
Add mime type to multipart parts

drklee3 committed:
use simd_json::ValueAccess

drklee3 committed:
cargo fmt yet again

drklee3 committed:
Pin mime dep to 0.3.16 for essence_str

drklee3 committed:
Another plural 'emojis' in intents

arqunis commented:
This has been manually merged: https://github.com/serenity-rs/serenity/commit/744bc1321a2c312b50da4c7ab1592944c6451168.
